### PR TITLE
Add support for fixing Semgrep's overly-permissioned rule

### PR DESCRIPTION
**Pull request by nahsra**

### CODE REVIEW

File: core-codemods/src/main/java/io/codemodder/codemods/SemgrepOverlyPermissiveFilePermissionsCodemod.java [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+      return false;
+    }
+  }
+
+  /**
+   * This handles the case where permissions are created from UNIX permission strings like:
+   *
+   * <p>{@code var p = Permissions.fromString("rwxrwxrwx")}
+   */
+  private static class FromStringChanger extends SarifPluginJavaParserChanger<ExpressionStmt> {
```
Comment (ryandens) [inline core-codemods/src/main/java/io/codemodder/codemods/SemgrepOverlyPermissiveFilePermissionsCodemod.java:89]: although this class could still only be subclassed in this file, marking this as `final` still helps static analysis tools provide better suggestions 
```suggestion
  private static final class FromStringChanger extends SarifPluginJavaParserChanger<ExpressionStmt> {
```

File: core-codemods/src/main/java/io/codemodder/codemods/SemgrepOverlyPermissiveFilePermissionsCodemod.java [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+      }
+    }
+    return false;
+  }
+
+  /**
+   * This handles the case where the permission is added to a set via {@link
+   * java.util.Set#add(Object)} call.
+   */
+  private static class PermissionAddCallChanger
```
Comment (ryandens) [inline core-codemods/src/main/java/io/codemodder/codemods/SemgrepOverlyPermissiveFilePermissionsCodemod.java:142]:
```suggestion
  private static final class PermissionAddCallChanger
```

File: core-codemods/src/main/java/io/codemodder/codemods/SemgrepOverlyPermissiveFilePermissionsCodemod.java [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+  }
+
+  /**
+   * This handles the case where the permission is added inline to {@code setPosixFilePermissions()}
+   * like:
+   *
+   * <p>{@code Files.setPosixFilePermissions(startupScript,
+   * PosixFilePermissions.fromString("rwxrwxrwx"));}
+   */
+  private static class InlineFromStringChanger
```
Comment (ryandens) [inline core-codemods/src/main/java/io/codemodder/codemods/SemgrepOverlyPermissiveFilePermissionsCodemod.java:40]:
```suggestion
  private static final class InlineFromStringChanger
```

File: framework/codemodder-base/src/main/java/io/codemodder/javaparser/ASTExpectations.java [hunk 4]
```diff
@@ -159,6 +217,13 @@ public MethodCallExpectation toBeInitializedByMethodCall() {
       return new MethodCallExpectation(mc);
     }
 
+    public LocalVariableDeclaratorExpectation withInitialization() {
```
Comment (ryandens) [inline framework/codemodder-base/src/main/java/io/codemodder/javaparser/ASTExpectations.java:220]: I don't see this used anywhere, which makes me confused how/why it should be used - should we delay adding this?

**Review by ryandens — APPROVED** [review]
minor feedback, but overall looks great to me! Excited to see this rule land

File: core-codemods/src/main/java/io/codemodder/codemods/SemgrepOverlyPermissiveFilePermissionsCodemod.java [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+public final class SemgrepOverlyPermissiveFilePermissionsCodemod
+    extends CompositeJavaParserChanger {
+
+  @Inject
+  public SemgrepOverlyPermissiveFilePermissionsCodemod(
+      final PermissionAddCallChanger addCallChanger,
+      final FromStringChanger fromStringChanger,
+      final InlineFromStringChanger inlineFromStringChanger) {
+    super(addCallChanger, fromStringChanger, inlineFromStringChanger);
+  }
```
[the remark this answers is not in the document]
Comment (nahsra) [reply core-codemods/src/main/java/io/codemodder/codemods/SemgrepOverlyPermissiveFilePermissionsCodemod.java:31]: I think I can do that in this instance!

**Pull request merged**